### PR TITLE
Introduce strict-text and posn-strict-text wrappers and monadUserState-strict-text

### PR DESCRIPTION
**Pull request by alexbiehl**
This PR introduces the strict-text and posn-strict-text wrappers to allows working with Data.Text out of the box.

Implementation wise we could do a bit better by using Data.Text.Foreign.takeWord8 and avoiding the intermediate list of bytes in a codepoint but this will do for now.

### CODE REVIEW

**Comment by alexbiehl** [comment]
Hm, tests do pass locally, not sure what's causing the issue.

**Comment by andreasabel** [comment]
`text` is only a GHC-shipped package from 8.4 on, compare the release notes:
- https://downloads.haskell.org/~ghc/8.2.2/docs/html/users_guide/8.2.1-notes.html
- https://downloads.haskell.org/~ghc/8.4.2/docs/html/users_guide/8.4.1-notes.html#included-libraries

Adding a dependency like `text` might require to cabalize the testsuite first.  (It's `Makefile` design is ancient, probably predating cabal.)

**Comment by alexbiehl** [comment]
@andreasabel It looks like to fully cabalize the test suite is a long term effort, how about I disable the text specific tests on old GHCs?

**Comment by andreasabel** [comment]
> how about I disable the text specific tests on old GHCs?

Yes, that seems fair, especially since you add a new feature so it is less important to test it with old GHCs.

File: doc/api.rst [hunk 1]
```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
 All of the actions in your lexical specification have the same type as in the ``monadUserState`` wrapper.
 It is only the types of the function to run the monad and the type of the ``token`` function that change.
 
+The text wrappers
+~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (andreasabel) [inline doc/api.rst:581]: `make -C doc` errors out with these warnings:
```
                                                                                 
/Users/abel/project/open-source/alex/doc/api.rst:623: WARNING: Title underline too short.

The "posn-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
api.rst:623: WARNING: Title underline too short.

The "posn-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
api.rst:650: WARNING: Title underline too short.

The "monad-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
api.rst:650: WARNING: Title underline too short.

The "monad-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
api.rst:689: WARNING: Title underline too short.

The "monadUserState-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
api.rst:689: WARNING: Title underline too short.

The "monadUserState-strict-text" wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: doc/api.rst [hunk 2]
```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
...
+     ( Char      -- previous char
+     , [Byte]    -- rest of the bytes for the current char
+     , Text      -- rest of the input Text
+     )
+
+   alexGetByte       :: AlexInput -> Maybe (Char, AlexInput)
+
+   alexInputPrevChar :: AlexInput -> Char
+
+   -- alexScanTokens :: ByteString -> [token]
```
Comment (andreasabel) [inline doc/api.rst:612]: C&P error? `ByteString`

File: doc/api.rst [hunk 3]
```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
...
+     }
+
+   newtype Alex a = Alex { unAlex :: AlexState
+                                  -> Either String (AlexState, a) }
+
+   instance Functor     Alex where ...
+   instance Applicative Alex where ...
+   instance Monad       Alex where ...
+
+   runAlex          :: Text -> Alex a -> Either String a
```
Comment (andreasabel) [inline doc/api.rst:674]: Please delete extra space after `runAlex`.

File: doc/api.rst [hunk 4]
```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
...
+     , alex_bytes :: [Byte]     -- rest of the bytes for the current char
+     , alex_scd   :: !Int       -- the current startcode
+     }
+
+   newtype Alex a = Alex { unAlex :: AlexState
+                                  -> Either String (AlexState, a) }
+
+   instance Functor     Alex where ...
+   instance Applicative Alex where ...
+   instance Monad       Alex where ...
```
Comment (andreasabel) [inline doc/api.rst:672]: Do we need these `instance` stubs?  
(In practice, one would use some automatic deriving tactics like `GeneralizedNewtypeDeriving` or `DerivingVia`...)

File: doc/api.rst [hunk 5]
```diff
@@ -577,6 +577,143 @@ The ``monadUserState-bytestring`` wrapper is the same as the ``monadUserState``
...
 It is only the types of the function to run the monad and the type of the ``token`` function that change.
 
+The text wrappers
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``strict-text``, ``posn-strict-text`` and ``monad-strict-text`` wrappers are variations on the ``basic``, ``posn`` and ``monad`` wrappers that use strict ``Text``\ s as the input and token types instead of an ordinary ``String``.
+
+The point of using these wrappers is that ``Text``\ s provide a more memory efficient representation of an input stream.
+They can also be somewhat faster to process.
+Note that using these wrappers adds a dependency on the ``Text`` modules, which live in the ``text`` package.
```
Comment (andreasabel) [inline doc/api.rst:587]: ...on the `Data.Text` modules

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for this new feature!

I have some small change requests concerning the documentation.

Unfortunately, the docs are not built automatically from the PR even though I switched it on  now at `readthedocs.org`.  Probably it only works on new PRs...

File: doc/api.rst (hunk 4 above)
Comment (alexbiehl) [reply doc/api.rst:672]: I think the piece of code is supposed to show users what the types are that are being generated, copied from above.

Comment (andreasabel) [reply doc/api.rst:672]: Ok, fair enough.  (Didn't see them for the `bytestring` wrapper, but nothing wrong having them here.)

**Review by andreasabel — APPROVED** [review]
I suppose we merge this, try it out a bit, and then release.

Ok to squash?
If you want to preserve separate commits, please reorganize them so that there are no fixups and each commit has a clear purpose and commit message.
Otherwise, we squash them into one big PR commit.

**Comment by alexbiehl** [comment]
Cool! Squash is fine.

**Comment by andreasabel** [comment]
I now tested this PR locally by porting one of my projects to the new `posn-strict-text` wrapper.  Worked!

**Comment by andreasabel** [comment]
A new major version (3.4.0.0) looks more appropriate for a new feature, one can then have  `build-tool-depends: alex:alex >= 3.4` (rather than `3.3.1`).

**Pull request merged**